src/client/Microsoft.Identity.Client/Instance/Region/RegionManager.cs

```diff
@@ -0,0 +1,242 @@
+            {
+                if (discoveredRegion.RegionSource != RegionSource.FailedAutoDiscovery)
+                {
+                    return discoveredRegion.Region;
+                }
+                else
+                {
+                    logger.Warning($"[Region discovery] Auto-discovery failed.");
+                    return null;
+                }
+            }
+
+            logger.Info($"[Region discovery] Returning user provided region {azureRegionConfig}");
+            return azureRegionConfig;
+        }
+
+        private static bool IsAutoDiscoveryRequested(string azureRegionConfig)
+        {
+            return string.Equals(azureRegionConfig, ConfidentialClientApplication.AttemptRegionDiscovery);
+        }
+
+        private void RecordTelemetry(ApiEvent apiEvent, string azureRegionConfig, RegionInfo discoveredRegion)
+        {
+            bool isAutoDiscoveryRequested = IsAutoDiscoveryRequested(azureRegionConfig);
+            apiEvent.RegionDiscovered = discoveredRegion.Region;
+
+            if (isAutoDiscoveryRequested)
+            {
+                apiEvent.RegionSource = (int)discoveredRegion.RegionSource;
+                if (discoveredRegion.Region != null)
+                {
+                    apiEvent.IsValidUserProvidedRegion = string.Equals(azureRegionConfig, discoveredRegion.Region);
+                }
+                else
+                {
+                    apiEvent.FallbackToGlobal = discoveredRegion.Region == null;
+                }
+            }
+            else
+            {
+                apiEvent.RegionSource = (int)RegionSource.UserProvided;
+                apiEvent.UserProvidedRegion = azureRegionConfig;
+            }
+        }
+
+        private async Task<RegionInfo> DiscoverRegionWithCacheAsync(string azureRegionConfig, ICoreLogger logger, CancellationToken requestCancellationToken)
+        {
+            if (s_failedAutoDiscovery == true)
+            {
+                logger.Info($"[Region discovery] Auto-discovery failed in the past. Not trying again.");
+                return new RegionInfo(null, RegionSource.FailedAutoDiscovery);
+            }
+
+            if (s_failedAutoDiscovery == false &&
+                !string.IsNullOrEmpty(s_autoDiscoveredRegion))
+            {
+                logger.Info($"[Region discovery] Auto-discovery already ran and found {s_autoDiscoveredRegion}");
+                return new RegionInfo(s_autoDiscoveredRegion, RegionSource.Cache);
+            }
+
+            var result = await DiscoverRegionNoCacheAsync(logger, requestCancellationToken).ConfigureAwait(false);
+
+            s_failedAutoDiscovery = result.RegionSource == RegionSource.FailedAutoDiscovery;
+            s_autoDiscoveredRegion = result.Region;
+
+            return result;
+        }
+
+        private async Task<RegionInfo> DiscoverRegionNoCacheAsync(ICoreLogger logger, CancellationToken requestCancellationToken)
```

Comment:
nit: not sure the "NoCache" is the best wording here. Perhaps the method on line #121 can be changed to "DiscoverAndCacheRegionAsync" and this method can just be "DiscoverRegionAsync"